### PR TITLE
Circular  issue #9

### PR DESCRIPTION
**Pull request by eino-makitalo**
Added stack handling so that routine detects groups already handled.  I reproduced problem in real Active Directory where user X  has membership to group A.

group A has member: group B
group B has member: group A

System went infinite loop and use all resources until stack exceeded.  It also connects several times to AD...

Special thanks to Joona Enbuska (Eficode) to help to find and fix this with me.

### CODE REVIEW

File: lib/components/getGroupMembershipForDN.js [hunk 1]
```diff
@@ -21,12 +21,12 @@ let search
...
  * @constructor
  */
-function GroupMembersForDN (opts, dn, callback) {
+function GroupMembersForDN (opts, dn, callback, stack) {
   this.opts = opts
   this.dn = dn
   this.callback = callback
-
-  this.stack = new Map()
+
```
Comment (jsumners) [inline lib/components/getGroupMembershipForDN.js]: Please remove this extraneous whitespace. Running `npm run lint` should catch such issues.

File: lib/components/getGroupMembershipForDN.js [hunk 2]
```diff
@@ -21,12 +21,12 @@ let search
...
  */
-function GroupMembersForDN (opts, dn, callback) {
+function GroupMembersForDN (opts, dn, callback, stack) {
   this.opts = opts
   this.dn = dn
   this.callback = callback
-
-  this.stack = new Map()
+  
+  this.stack = stack || new Map();
```
Comment (jsumners) [inline lib/components/getGroupMembershipForDN.js]: Please remove the trailing semi-colon (`npm run lint`).

**Comment by eino-makitalo** [comment]
Yes that was accidental that package.lock file..  I check next week this.

**Comment by jsumners** [comment]
@eino-makitalo I want to incorporate your changes. Are you able to address the feedback I have given?

**Comment by eino-makitalo** [comment]
I have to check that testing issue. How you have mocked Active directory queries but basic idea is to simulate isituation where User X is member of group A.  Group A is member of B and B is member of A. (this makes it stack overflow without this fix

File: lib/components/getGroupMembershipForDN.js (hunk 1 above)
Comment (eino-makitalo) [reply lib/components/getGroupMembershipForDN.js]: My eslint was not saying nothing but I fixed it so that npm run lint is happy aka silent.

**Comment by jsumners** [comment]
The basic gist of it is I had to write an LDAP server with ldapjs, but I had to override some stuff in ldapjs because of the various authorization mechanisms AD supports that vanilla LDAP doesn't. It's a mess. Not being able to add a test is not a blocker for the PR being accepted. It would be in almost every other project, but I can't blame anyone for not wanting to deal with the headache needed to make sure AD works right.

**Comment by eino-makitalo** [comment]
Ok I added person Circular Celebrity which tests this previous bug. I'm also happy to inform that problems in real life stopped after this little fix.

**Comment by jsumners** [comment]
Excellent. Thank you very much for this PR.

**Pull request merged**